### ISSUE TITLE
Not able to reproduce the validation results of the hyperparameter sweep in PyTorch Lightning

### ISSUE THREAD

**Issue by Adi-iitd** [wandb/examples#72]
Hi, I am not able to reproduce the training and validation results from the model trained separately on the best hyperparameter chosen from the wandb sweep. 

The reason being, I am only able to reproduce the result of the search corresponding to the first param config (if a separate model is trained for this hyperparameter config). From the second config of the sweep, I am not able to reproduce the exact same results if a separate model is trained for this config.

I just want to double-check if I am writing the sweep_iteration function correctly or not. Code is modular and you can skim through it very quickly. [COLAB LINK](https://colab.research.google.com/drive/15AG16xkwNOnsk7aorCHUx_pEFKgFS8ZE?usp=sharing)

`
class T5Dataset(pl.LightningDataModule):
    
    def __init__(self, args: argparse.Namespace):
        
        super().__init__()
        self.data_dir = args.data_dir
        
        self.max_src_length = args.max_src_length
        self.max_tgt_length = args.max_tgt_length
        
        self.train_batch_sz = args.train_batch_sz
        self.eval_batch_sz  = args.eval_batch_sz
        
        if args.do_train:
            config = AutoConfig.from_pretrained(args.model_name)
            self.tokenizer = AutoTokenizer.from_pretrained(args.model_name, config = config)
        
        if args.do_predict:
            self.tokenizer = AutoTokenizer.from_pretrained(args.output_dir)
        
    
    def prepare_data(self):
        pass
        
    
    def setup(self, stage: str = None):
        
        if stage == "fit"  or stage == None:
            self.train = MyDataset(data_dir = self.data_dir, 
                                   data_type = "train", 
                                   tokenizer = self.tokenizer,
                                   max_src_length = self.max_src_length, 
                                   max_tgt_length = self.max_tgt_length)
            
            self.valid = MyDataset(data_dir = self.data_dir, 
                                   data_type = "valid", 
                                   tokenizer = self.tokenizer, 
                                   max_src_length = self.max_src_length, 
                                   max_tgt_length = self.max_tgt_length)
            
            print(f"Length of train dataset: {len(self.train)}, valid dataset: {len(self.valid)}")
        
        if stage == "test" or stage == None:
            self.test  = MyDataset(data_dir = self.data_dir, 
                                   data_type = "test",
                                   tokenizer = self.tokenizer, 
                                   max_src_length = self.max_src_length, 
                                   max_tgt_length = self.max_tgt_length)
            
            print(f"Length of the test dataset: {len(self.test)}")
            
    
    def train_dataloader(self):
        return DataLoader(self.train, self.train_batch_sz, shuffle = True, num_workers = 16)
    
    def val_dataloader(self):
        return DataLoader(self.valid, self.eval_batch_sz, shuffle = False, num_workers = 16)
    
    def test_dataloader(self):
        return DataLoader(self.test , self.eval_batch_sz, shuffle = False, num_workers = 16)

****************************************************************************************

class T5Model(pl.LightningModule):
    
    def __init__(self, args: argparse.Namespace):
        
        super().__init__()
        self.hparams = args
        self.save_hyperparameters()
        
        self.dataset = T5Dataset(self.hparams)
        self.dataset.prepare_data()
        
        if args.do_train:
            
            self.dataset.setup("fit")
            config = AutoConfig.from_pretrained(self.hparams.model_name)
            
            self.tokenizer = AutoTokenizer.from_pretrained(self.hparams.model_name)
            self.model = AutoModelForSeq2SeqLM.from_pretrained(self.hparams.model_name, config = config)
            
        if args.do_predict:
            
            self.dataset.setup("fit")
            self.dataset.setup("test")
            
            self.tokenizer = AutoTokenizer.from_pretrained(self.hparams.output_dir)
            self.model = AutoModelForSeq2SeqLM.from_pretrained(self.hparams.output_dir)
            
    
    def configure_optimizers(self):
        
        optimizer = Adafactor(self.model.parameters(), 
                              lr = self.hparams.learning_rate,
                              relative_step = False, 
                              scale_parameter = False)
        self.opt = optimizer
        return [optimizer]
        
        
    def forward(self, input_ids, attention_mask = None, labels = None):
        
        return self.model(input_ids = input_ids, attention_mask = attention_mask, labels = labels)
    
    
    def _step(self, batch, return_text: bool = False):
        
        input_ids, src_mask, target_ids = batch["source_ids"], batch["source_mask"], batch["target_ids"]
        target_ids_copy = target_ids.clone()
        
        target_ids[target_ids[:, :] == self.tokenizer.pad_token_id] = -100
        outputs = self(input_ids = input_ids, attention_mask = src_mask, labels = target_ids)
        loss = outputs[0]
        
        if return_text:
            pred_ids = [self.tokenizer.decode(ids,
                                              skip_special_tokens = True, 
                                              clean_up_tokenization_spaces = True) for ids in target_ids_copy]
            return loss, pred_ids
        
        return loss
    
    
    def training_step(self, batch, batch_idx):
        
        loss = self._step(batch, return_text = False)
        
        self.log("train_loss", loss, on_step = True, on_epoch = True, prog_bar = True, logger = True)
        return loss
    
    
    def validation_step(self, batch, batch_idx):
        
        loss, true_text = self._step(batch, return_text = True)
        pred_text = self.test_step(batch, batch_idx)["pred"]
        
        f1_score = f1(true_text, pred_text)
        exact_match_score = exact(true_text, pred_text)
        
        dict_ = {"val_loss": loss, "f1_score": f1_score, "exact_match_score": exact_match_score}
        self.log_dict(dict_, on_step = False, on_epoch = True, prog_bar = True, logger = True)
        
        return dict_
    
    
    def test_step(self, batch, batch_idx):
        
        input_ids, src_mask = batch["source_ids"], batch["source_mask"]
        pred_ids = self.model.generate(
                                    num_beams = 1,
                                    do_sample = False,
                                    early_stopping = True,
                                    num_return_sequences = 1,
                                    input_ids = input_ids, 
                                    attention_mask = src_mask,
                                    top_k = self.hparams.top_k,
                                    top_p = self.hparams.top_p,
                                    max_length = self.hparams.max_length,
                                    length_penalty = self.hparams.length_penalty,
                                    repetition_penalty = self.hparams.repetition_penalty)

        pred = [self.tokenizer.decode(ids, 
                                      skip_special_tokens = True, 
                                      clean_up_tokenization_spaces = True) for ids in pred_ids]
        return {"pred": pred}
    
    
    def test_epoch_end(self, outputs):
        
        comp_pred = []
        output_file = os.path.join(self.hparams.output_dir, "test_predictions.txt")
        
        with open(output_file, "w+") as file:
            for out in outputs:
                file.writelines(s + "\n" for s in out["pred"])
                comp_pred += out["pred"]
        
        return {"pred": comp_pred}

****************************************************************************************

def add_args(parser: argparse.Namespace, hyperparam_opt_flag: bool = False):
    
    if not hyperparam_opt_flag:
        
        parser.add_argument("--logger", default = wandb_logger)
        
        parser.add_argument("--output_dir", default = "./Outputs/", type = str,
                            help = "Output directory where model predictions and checkpoints will be stored.")
        parser.add_argument("--pretrained_model_path", default = None, type = str,
                            help = "The output directory where the pre-trained model is saved.")
    
    parser.add_argument("--data_dir", default = "./Dataset/", type = str)
    parser.add_argument("--model_name", default = "t5-base", type = str)
    
    parser.add_argument("--max_epochs", default = 1, type = int)
    parser.add_argument("--learning_rate", default = 1e-3, type = float)
    
    parser.add_argument("--top_k", default = 50, type = int)
    parser.add_argument("--top_p", default = 0.95, type = float)
    
    parser.add_argument("--max_src_length", default = 128, type = int)
    parser.add_argument("--max_tgt_length", default = 64, type = int)
    parser.add_argument("--train_batch_sz", default = 32, type = int)
    parser.add_argument("--eval_batch_sz" , default = 32, type = int)
    
    parser.add_argument("--max_length", default = 64, type = int)
    parser.add_argument("--length_penalty", default = 2.0, type = float)
    parser.add_argument("--repetition_penalty", default = 1.0, type = float)
    
    parser.add_argument("--max_grad_norm", default = 1.0, type = float)
    parser.add_argument("--gradient_accumulation_steps", default = 1, type = int,
                        help = "Number of update steps to accumulate before performing a backward pass.")
    
    parser.add_argument("--n_gpu", default = 1, type = int)
    parser.add_argument("--fp16", action = "store_true", 
                        help = "Whether to use 16-bit (mixed) precision instead of 32-bit")
    
    parser.add_argument("--do_train", action = "store_true")
    parser.add_argument("--do_predict", action = "store_true")

****************************************************************************************

  sweep_config = {
  
  "method": "bayes",
  "metric": {
      "name": "f1_score",
      "goal": "maximize",
      "target": 0.90
  },  
  "parameters": {
        "learning_rate": {
            "distribution": "log_uniform",
            "min": -9.21,
            "max": -4.61
        },
        "gradient_accumulation_steps": {
            "values": [2, 3, 4]
        },
    }
  }

****************************************************************************************

def sweep_iteration():
    
    wandb.init()
    wandb_logger = WandbLogger()
    
    ARGS = f"""
    --train_batch_sz={str(wandb.config.train_batch_sz)} \
    --learning_rate={str(wandb.config.learning_rate)} \
    --max_epochs={str(wandb.config.max_epochs)} \
    --gradient_accumulation_steps={str(wandb.config.gradient_accumulation_steps)}
    --do_train
    """
    
    parser = argparse.ArgumentParser()
    add_args(parser, hyperparam_opt_flag = True)
    args = parser.parse_args(ARGS.split())
    
    early_stop_callback = pl.callbacks.EarlyStopping(monitor = 'f1_score', 
                                                     min_delta = 0.0, 
                                                     patience = 3,
                                                     verbose = False,
                                                     mode = 'max')
    
    t5 = T5Model(args)
    
    train_params = dict(
                        gpus = args.n_gpu,
                        logger = wandb_logger,
                        max_epochs = wandb.config.max_epochs,
                        callbacks = [early_stop_callback],
                        gradient_clip_val = args.max_grad_norm,
                        accumulate_grad_batches = wandb.config.gradient_accumulation_steps)
    
    train_params["deterministic"] = True
    if args.fp16: train_params["precision"] = 16
    if args.n_gpu > 1: train_params["distributed_backend"] = "ddp"
    
    trainer = pl.Trainer(**train_params)
    trainer.fit(t5, t5.dataset)

    wandb.finish()

os.environ['WANDB_WATCH'] = 'false'
wandb.agent(sweep_id, function = sweep_iteration)
`

**Comment by charlesfrye** [comment]
Hey there:

We use the issues section here only for issues with the examples in this repository, rather than with the W&B client library or its integrations. If you believe this issue to be caused by W&B, I would suggest raising an issue in the [client library repo](http://github.com/wandb/client).

To do so, you'll want to write a minimal example in which the library behaves incorrectly. For example, start with just initiating a sweep that prints the config and a random number (after setting your seeds). If that works -- the random number is the same -- add more functionality from your code until it stops working. I suspect that the issue might be caused by pRNG seeds not being set completely (try using pytorch lightning's [`seed_everything`](https://pytorch-lightning.readthedocs.io/en/latest/api/pytorch_lightning.utilities.seed.html#pytorch_lightning.utilities.seed.seed_everything) instead of trying to roll your own).
